pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +380,565 @@ pub mod pallet {
+			investment_id: T::InvestmentId,
+			_foreign_payout_currency: T::CurrencyId,
+			_pool_currency: T::CurrencyId,
+		) -> DispatchResult {
+			T::Investment::collect_redemption(who.clone(), investment_id)
+		}
+
+		fn investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::investment(who, investment_id)
+		}
+
+		fn redemption(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> Result<T::Balance, DispatchError> {
+			T::Investment::redemption(who, investment_id)
+		}
+
+		fn accepted_payment_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payment_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(pool_currency, currency))
+					.unwrap_or(false)
+			}
+		}
+
+		fn accepted_payout_currency(
+			investment_id: T::InvestmentId,
+			currency: T::CurrencyId,
+		) -> bool {
+			if T::Investment::accepted_payout_currency(investment_id, currency) {
+				true
+			} else {
+				T::PoolInspect::currency_for(investment_id.of_pool())
+					.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+					.unwrap_or(false)
+			}
+		}
+	}
+
+	pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+		type Error = DispatchError;
+		type Id = T::SwapId;
+		type Status = SwapOf<T>;
+
+		fn notify_status_change(
+			swap_id: T::SwapId,
+			last_swap: SwapOf<T>,
+		) -> Result<(), DispatchError> {
+			let (who, investment_id) =
+				SwapidToForeignId::<T>::get(swap_id).ok_or(Error::<T>::SwapOrderNotFound)?;
+
+			let remaining_swap_amount_in = match T::TokenSwaps::get_order_details(swap_id) {
+				Some(swap) => swap.amount_in,
+				None => {
+					Pallet::<T>::unregister_swap(&who, investment_id, swap_id);
+					T::Balance::zero()
+				}
+			};
+
+			Pallet::<T>::notify_swap_done(
+				&who,
+				investment_id,
+				last_swap.currency_out,
+				last_swap.amount_in,
+				remaining_swap_amount_in,
+			)
+		}
+	}
+
+	pub struct CollectedInvestmentHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for CollectedInvestmentHook<T> {
+		type Error = DispatchError;
+		type Id = (T::AccountId, T::InvestmentId);
+		type Status = CollectedAmount<T::Balance>;
+
+		fn notify_status_change(
+			(who, investment_id): (T::AccountId, T::InvestmentId),
+			collected: CollectedAmount<T::Balance>,
+		) -> DispatchResult {
+			let info = ForeignInvestmentInfo::<T>::mutate(&who, investment_id, |maybe_info| {
+				let info = maybe_info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+				info.base.collected_amount.increase(&collected)?;
+				Ok::<_, DispatchError>(info.clone())
+			})?;
+
+			if info.remaining_foreign_amount()?.is_zero() {
+				ForeignInvestmentInfo::<T>::remove(&who, investment_id);
+			}
```

Comment:
See my above remark that `remaining_pool_amount` should be used instead. Moreover, this check might be redundant because we increase `collected_amount` by `collected` such that the remaining amount can only be zero if it was zero and incremented by zero (0 + 0).

Reply:
You're right, the amounts mismatch here. I'll think for a clean solution here